factor out code from pkg/endpoint, cilium-health which depends on pkg/k8s

ianvernon opened the pull request:
This PR does some code reorganizing to inject the Kubernetes dependency into the endpoint package and the cilium-health package at runtime. 

I had to change `endpoint.getLogger()` to be exported to allow for this, as well as expose the capability of updating a controller for an endpoint by external packages. I haven't done dependency injection before this, so I am open to alternative proposals in how to remove these dependencies.

Signed-off by: Ian Vernon <ian@cilium.io>

Related-to: #6101

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6126)
<!-- Reviewable:end -->

Commit:
k8s: move node annotation functions to annotate.go

No functional changes are intended by this commit.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/endpoint/kubernetes/cep.go:
```diff
@@ -0,0 +1,348 @@
+package kubernetes
```
Missing copyright header.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -485,6 +314,12 @@ type Endpoint struct {
 
+// UpdateController updates the controller with the specified name with the
+// provided list of parameters in endpoint's list of controllers.
+func (e *Endpoint) UpdateController(name string, params controller.ControllerParams) *controller.Controller {
```
This functions seems redundant.

ianvernon replied:
The controllers for the endpoint are not exported. This is done so consumers of endpoints can update controllers for it.

aanm commented on pkg/endpoint/kubernetes/cep.go:
```diff
@@ -0,0 +1,348 @@
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/sirupsen/logrus"
+	"reflect"
```
random order in imports

aanm commented on pkg/k8s/annotate.go:
```diff
@@ -16,38 +16,26 @@ package k8s
 
 import (
 	"fmt"
+	"k8s.io/client-go/kubernetes"
```
Import order.

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -221,6 +221,12 @@ func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressin
 	ep.SetDefaultOpts(option.Config.Opts)
 
+	if k8s.IsEnabled() {
+		ep.Annotator = k8s.K8sCli
```
cilium-health is won't be aware if k8s is or is not enabled.

joestringer replied:
This should be happening from the launcher side, in the daemon no?

ianvernon replied:
You are right. I've marked this as wip to fix up these concerns.

aanm commented on pkg/k8s/client.go:
```diff
@@ -40,6 +40,8 @@ var (
 	// ErrNilNode is returned when the Kubernetes API server has returned a nil node
 	ErrNilNode = goerrors.New("API server returned nil node")
 
+	K8sCli = &K8sClient{}
```
Do something like this instead.
```patch
diff --git a/pkg/k8s/client.go b/pkg/k8s/client.go
index 0076c5b2d..d1e5138eb 100644
--- a/pkg/k8s/client.go
+++ b/pkg/k8s/client.go
@@ -42,7 +42,7 @@ var (
 
        // client is the object through which interactions with Kubernetes are
        // performed.
-       client kubernetes.Interface
+       client *OurClient
 )
 
 // CreateConfig creates a rest.Config for a given endpoint using a kubeconfig file.
@@ -223,11 +223,19 @@ func AnnotateNode(c kubernetes.Interface, nodeName string, v4CIDR, v6CIDR *net.I
        return nil
 }
 
+type OurClient struct {
+       kubernetes.Interface
+}
+
 // Client returns the default Kubernetes client
-func Client() kubernetes.Interface {
+func Client() *OurClient {
        return client
 }
 
+func (c OurClient) AnnotatePod(k8sNamespace, k8sPodName, annotationKey, annotationValue string) error {
+       return nil
+}
+
 func createDefaultClient() error {
        restConfig, err := CreateConfig()
        if err != nil {
@@ -239,7 +247,9 @@ func createDefaultClient() error {
                return fmt.Errorf("unable to create k8s client: %s", err)
        }
 
-       client = k8sClient
+       client = &OurClient{
+               Interface: k8sClient,
+       }
 
        return nil
 }
```

in the endpoint package:
```
	if k8s.IsEnabled() {
		ep.Annotator = k8s.Client()
	}
```

ianvernon replied:
I went ahead and did something just like this. Thanks!

aanm commented on daemon/endpoint.go:
```diff
@@ -141,6 +142,11 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
 		return PutEndpointIDInvalidCode, err
 	}
 	ep.SetDefaultOpts(option.Config.Opts)
+	if k8s.IsEnabled() {
```
Do this assignment inside `NewEndpointFromChangeModel`

ianvernon replied:
I disagree with this because the annotator is not provided by the change model; the annotator is set from daemon configuration. Endpoint configuration is set above as well. I can factor this logic to set the annotator based off of k8s being enabled or not as a helper function and call that here:

```
func getEndpointAnnotator() endpoint.Annotator {
	var annotator endpoint.Annotator
	if k8s.IsEnabled() {
		annotator = k8s.K8sCli
	} else {
		annotator = k8s.DummyAnnotator{}
	}
	return annotator
}
```

Would that address your concerns?

ianvernon replied:
This is resolved by no longer having the Annotator stored within the Endpoint structure.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -379,11 +379,11 @@ func (e *Endpoint) addNewRedirects(owner Owner, m *policy.L4Policy, proxyWaitGro
 
 	return desiredRedirects, nil, finalizeList.Finalize, func() error {
-		e.getLogger().Debug("Reverting proxy redirect additions")
+		e.GetLogger().Debug("Reverting proxy redirect additions")
```
Specify "why?" for this change in commit message.

aanm commented on pkg/k8s/annotate.go:
```diff
@@ -16,9 +16,17 @@ package k8s
+	"net"
+	"time"
 
+	"github.com/sirupsen/logrus"
```
order:
- lib
- cilium
- others

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -18,6 +18,7 @@ import (
 	"bufio"
 	"context"
 	"fmt"
+	"net"
```
The commit messages doesn't describe the actual changes.
> cilium-health: remove direct dependency on pkg/k8s

How can this be possible if there is no import removal of a `pkg/k8s`?

ianvernon replied:
I goofed and marked this as pending-review when it clearly shouldn't have been. Will test and request review if tests pass.

Commit:
use dependency injection for K8s dependencies

pkg/endpoint, pkg/endpointmanager, and cilium-health/launcher all utilized
pkg/k8s directly. This is not ideal, as importing pkg/k8s pulls in all of the
transitive dependencies of Kubernetes (around 120 dependencies). The
aforementioned packages  only require interaction with Kubernetes in a few,
isolated ways, specifically updating resources (pods, nodes, and
CiliumEndpoints). Have these packages define interfaces which can then be
implemented outside of said package, and inject the Kubernetes dependency via
the daemon into these packages.

Removing the need for pkg/k8s to be imported resulted in a decrease of about
36 % in dependencies for all three packages.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -297,6 +125,7 @@ type PolicyMapStateEntry struct {
 // is not JSON forward compatible.
 //
 type Endpoint struct {
+	Annotator Annotator `json:"-"`
```
Add comment. Now that I'm thinking of, do we need to have an annotator per endpoint? This can be shared across all of them.

ianvernon replied:
Good idea, that's much cleaner!

aanm commented on pkg/k8s/client.go:
```diff
@@ -134,9 +133,14 @@ func isConnReady(c *kubernetes.Clientset) error {
+}
+
+// K8sCli returns the default K8sClient object.
+func K8sCli() *K8sClient {
```
You can remove this funtion and use `Client()` directly.

aanm commented on pkg/k8s/client.go:
```diff
@@ -134,9 +133,14 @@ func isConnReady(c *kubernetes.Clientset) error {
 
-// Client returns the default Kubernetes client
+// Client returns the default Kubernetes client.
 func Client() kubernetes.Interface {
```
This would return `*K8sClient`

aanm commented on pkg/k8s/annotate.go:
```diff
@@ -30,30 +29,34 @@ import (
-	// GetK8sPodName returns the name of the pod
-	GetK8sPodName() string
+// DummyAnnotator is a placeholder annotator which does nothing.
+type DummyAnnotator struct {
```
You can remove everything from lines 39-53

aanm commented on pkg/endpoint/kubernetes/cep.go:
```diff
@@ -0,0 +1,367 @@
+// Copyright 2016-2018 Authors of Cilium
```
what if we move `pkg/endpoint/kubernetes` to `pkg/k8s/cilium_endpoint` instead?

ianvernon replied:
I've done this. The only drawback here is that this puts the endpoint dependency into `pkg/k8s`.

aanm commented on daemon/endpoint.go:
```diff
@@ -132,14 +134,32 @@ func NewPutEndpointIDHandler(d *Daemon) PutEndpointIDHandler {
+func getEndpointAnnotator() endpoint.Annotator {
+	var annotator endpoint.Annotator
+	if k8s.IsEnabled() {
+		annotator = k8s.Client()
```
Don't need to create the variable you can directly return `nil` or `k8s.Client()`

ianvernon replied:
Will be removed in next upload.

aanm commented on daemon/endpoint.go:
```diff
@@ -132,14 +134,32 @@ func NewPutEndpointIDHandler(d *Daemon) PutEndpointIDHandler {
+func getHealthAnnotator() launch.Annotator {
+	var annotator launch.Annotator
+	if k8s.IsEnabled() {
+		annotator = k8s.Client()
```
Don't need to create the variable you can directly return `nil` or `k8s.Client()`

ianvernon replied:
I'll remove these functions in the next upload.

aanm commented on daemon/main.go:
```diff
@@ -966,11 +969,17 @@ func runDaemon() {
 	}
+
+	// Inject K8s dependency into packages which need to annotate K8s resources.
+	endpoint.EpAnnotator = getEndpointAnnotator()
```
did you try `endpoint.EpAnnotator = k8s.Client()` 4 lines above? (inside `k8s.IsEnabled()`)

ianvernon replied:
Good idea - I'll remove the added helper functions in the next upload.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -16,6 +16,7 @@ package endpoint
 
 import (
 	"fmt"
+	"github.com/cilium/cilium/pkg/uuid"
```
import order

ianvernon replied:
Sigh, sorry for the noise. I think I messed up my automatic goimports format in my IDE...

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -889,8 +889,23 @@ func (e *Endpoint) runIdentityToK8sPodSync() {
+				// checking endpoint state instead.
 				if id != "" && e.GetK8sNamespace() != "" && e.GetK8sPodName() != "" {
-					return k8s.AnnotatePod(e, k8sConst.CiliumIdentityAnnotation, id)
+					if EpAnnotator != nil {
```
this shouldn't be necessary as `EpAnnotator` is only set if k8s is enabled and by the name of the function `runIdentityToK8sPodSync` I'm guessing it's only going to be executed when k8s is enabled.

ianvernon replied:
I'd rather have the check than avoid a panic.

aanm replied:
Actually there's a bug in the current code. Please put
```
if !k8s.Enabled() {
     return nil
}
```
on line 879. We shouldn't even run a controller if we are not in k8s mode.

I know what you're going to say, the answer is:

If you want to avoid using k8s dependency in this package then `runIdentityToK8sPodSync` should be in the k8s package the same way `RunK8sCiliumEndpointSync` is. (If you do this do it in a new commit).

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -695,7 +694,7 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
 	d.k8sAPIGroups.addAPI(k8sAPIGroupNamespaceV1Core)
 
-	endpoint.RunK8sCiliumEndpointSyncGC()
+	k8s.RunK8sCiliumEndpointSyncGC()
```
k8s.RunK8s stutters a bit, but not a biggie.

ianvernon replied:
Good catch, will fix in next upload.

joestringer commented on daemon/main.go:
```diff
@@ -966,11 +969,17 @@ func runDaemon() {
 	// Launch another cilium-health as an endpoint, managed by cilium.
 	log.Info("Launching Cilium health endpoint")
 	if k8s.IsEnabled() {
+
```
extraneous newline?

ianvernon replied:
Removed in next upload.

joestringer commented on pkg/k8s/cep.go:
```diff
@@ -0,0 +1,363 @@
+	var (
+		endpointID     = e.ID
+		controllerName = fmt.Sprintf("sync-to-k8s-ciliumendpoint (%v)", endpointID)
+		scopedLog      = e.GetLogger().WithField("controller", controllerName)
```
Please use `e.Logger(subsystem)` instead. This avoids the need for exporting `getLogger()`, and provides the context of where the log message actually came from rather than misleading that it comes from the endpoint package.

ianvernon replied:
Ah, I didn't know that function existed. Dropped the `getLogger` --> `GetLogger` commit in next upload and changed this function to use `Logger`.

joestringer reviewed (changes requested):
I have a couple of minor comments, one of them would allow you to not have to export `getLogger()` just by using the existing `Logger()` function.

joestringer commented on pkg/k8s/cep.go:
```diff
@@ -0,0 +1,363 @@
+
+// EndpointSynchronizer currently is an empty type, which wraps around syncing
+// of CiliumEndpoint resources.
+// TODO - see whether folding the global variables below into this function
```
Has this TODO been investigated?

ianvernon replied:
Not yet - I think that is the next logical step that can be achieved by a follow-up commit.

joestringer commented on pkg/k8s/annotate.go:
```diff
@@ -16,38 +16,31 @@ package k8s
-	StringID() string
+// K8sClient is a wrapper around kubernetes.Interface.
+type K8sClient struct {
+	// client is the object through which interactions with Kubernetes are
```
Outdated comment referring to a 'client' field?

ianvernon replied:
Will fix in next upload.

joestringer commented on pkg/k8s/cep.go:
```diff
@@ -0,0 +1,363 @@
+				case err != nil && k8serrors.IsNotFound(err):
+					break
+
+					// Delete the CEP on the first ever run. We will fall through to the create code below
```
The indentation for comments above these switch cases are a bit inconsistent (compare with line 289).

ianvernon replied:
Will fix in next upload.

joestringer reviewed (approved):
Some minor tidy up comments below, but otherwise LGTM.
